File: src/api/descriptors.ts
import { invoke } from "./client";
import type {
  DescriptorField,
  DescriptorFieldPatch,
  DescriptorMatrix,
  DescriptorValue,
  NewDescriptorField,
} from "./types";

export const listDescriptorFields = () => invoke<DescriptorField[]>("list_descriptor_fields");
export const createDescriptorField = (input: NewDescriptorField) =>
  invoke<DescriptorField>("create_descriptor_field", { input });
export const updateDescriptorField = (id: string, patch: DescriptorFieldPatch) =>
  invoke<DescriptorField>("update_descriptor_field", { id, patch });
export const deleteDescriptorField = (id: string) =>
  invoke<DescriptorField>("delete_descriptor_field", { id });
export const reorderDescriptorFields = (ids: string[]) =>
  invoke<DescriptorField[]>("reorder_descriptor_fields", { ids });
export const setDescriptorValue = (documentId: string, fieldId: string, value: string | null) =>
  invoke<DescriptorValue | null>("set_descriptor_value", { documentId, fieldId, value });
export const listDocumentDescriptorValues = (documentId: string) =>
  invoke<DescriptorValue[]>("list_document_descriptor_values", { documentId });
export const getDescriptorMatrix = () => invoke<DescriptorMatrix>("get_descriptor_matrix");

File: src-tauri/src/commands/descriptors.rs
use misket_core::db::descriptors;
use misket_core::models::{
    DescriptorField, DescriptorFieldPatch, DescriptorMatrix, DescriptorValue, NewDescriptorField,
};
use misket_core::Result;
use tauri::State;

use crate::state::AppState;

#[tauri::command]
pub fn list_descriptor_fields(state: State<'_, AppState>) -> Result<Vec<DescriptorField>> {
    state.with_project(|p| descriptors::list_fields(&p.conn))
}

#[tauri::command]
pub fn create_descriptor_field(
    state: State<'_, AppState>,
    input: NewDescriptorField,
) -> Result<DescriptorField> {
    state.with_project(|p| descriptors::create_field(&p.conn, input))
}

#[tauri::command]
pub fn update_descriptor_field(
    state: State<'_, AppState>,
    id: String,
    patch: DescriptorFieldPatch,
) -> Result<DescriptorField> {
    state.with_project(|p| descriptors::update_field(&p.conn, &id, patch))
}

#[tauri::command]
pub fn delete_descriptor_field(state: State<'_, AppState>, id: String) -> Result<DescriptorField> {
    state.with_project(|p| descriptors::delete_field(&p.conn, &id))
}

#[tauri::command]
pub fn reorder_descriptor_fields(
    state: State<'_, AppState>,
    ids: Vec<String>,
) -> Result<Vec<DescriptorField>> {
    state.with_project(|p| descriptors::reorder_fields(&p.conn, &ids))
}

#[tauri::command]
pub fn set_descriptor_value(
    state: State<'_, AppState>,
    document_id: String,
    field_id: String,
    value: Option<String>,
) -> Result<Option<DescriptorValue>> {
    state.with_project(|p| {
        descriptors::set_value(&p.conn, &document_id, &field_id, value.as_deref())
    })
}

#[tauri::command]
pub fn list_document_descriptor_values(
    state: State<'_, AppState>,
    document_id: String,
) -> Result<Vec<DescriptorValue>> {
    state.with_project(|p| descriptors::values_for_document(&p.conn, &document_id))
}

#[tauri::command]
pub fn get_descriptor_matrix(state: State<'_, AppState>) -> Result<DescriptorMatrix> {
    state.with_project(|p| descriptors::values_matrix(&p.conn))
}
